### ISSUE TITLE
[bug] logits processor race condition in overlap mode

### ISSUE THREAD

**Issue by akhoroshev** [sgl-project/sglang#8056]
[One](https://github.com/sgl-project/sglang/blob/4a8837950abb7a39d5b890b8f4ee21bd9ded959d/test/srt/test_srt_endpoint.py#L452) of the logit processor tests is disabled because we have race condition on output_ids variable.
```python
        """
        NOTE: This feature has a race condition bug.
        This line https://github.com/sgl-project/sglang/blob/ef8ec07b2ce4c70c2a33ec5acda4ce529bc3cda4/test/srt/test_srt_endpoint.py#L395-L396 can be accessed by two concurrent threads at the same time. The access order is not guaranteed.
        In sglang, we use two python threads to overlap the GPU computation and CPU scheduling.
        Thread 1 (the CPU scheduling thread) will update the `param_dict["__req__"].output_ids`.
        Thread 2 (the GPU computation thread) will call `DeterministicStatefulLogitProcessor` because sampling is considered as GPU computation.
        We can fix this by moving the call of DeterministicStatefulLogitProcessor to the CPU scheduling thread.
        """
```

I guess we also have race condition on [custom_params](https://github.com/sgl-project/sglang/blob/4a8837950abb7a39d5b890b8f4ee21bd9ded959d/python/sglang/srt/sampling/sampling_batch_info.py#L57) and [custom_logit_processor](https://github.com/sgl-project/sglang/blob/4a8837950abb7a39d5b890b8f4ee21bd9ded959d/python/sglang/srt/sampling/sampling_batch_info.py#L59C5-L59C27) variables since CPU scheduling thread will update them in filter_batch/merge_batch functions while GPU computation thread will access them in apply_custom_logit_processor function.

**Comment by akhoroshev** [comment]
@merrymercy

**Comment by frrad** [comment]
I'm seeing behavior which could be explained by this bug. (Specifically, I have a stateful `custom_logit_processor` which is basically a state machine, reaching states which should be unreachable.)

Is there any recommended workaround?

cc @zhaochenyang20 if you have any suggestions

**Comment by zhaochenyang20** [comment]
> I'm seeing behavior which could be explained by this bug. (Specifically, I have a stateful `custom_logit_processor` which is basically a state machine, reaching states which should be unreachable.)
> 
> Is there any recommended workaround?
> 
> cc [@zhaochenyang20](https://github.com/zhaochenyang20) if you have any suggestions

Hey, quite amazed me. thanks @frrad @akhoroshev 

Let me bring this to our team.

**Comment by Edenzzzz** [comment]
> I guess we also have race condition on [custom_params](https://github.com/sgl-project/sglang/blob/4a8837950abb7a39d5b890b8f4ee21bd9ded959d/python/sglang/srt/sampling/sampling_batch_info.py#L57) and [custom_logit_processor](https://github.com/sgl-project/sglang/blob/4a8837950abb7a39d5b890b8f4ee21bd9ded959d/python/sglang/srt/sampling/sampling_batch_info.py#L59C5-L59C27) variables since CPU scheduling thread will update them in filter_batch/merge_batch functions while GPU computation thread will access them in apply_custom_logit_processor function.

I don't think the output_ids in [merge_batch](https://github.com/sgl-project/sglang/blob/004f7f1972dcec9dd408e65e31e9e68bc79c4940/python/sglang/srt/managers/schedule_batch.py#L1676) has a race condition with GPU comp--the [output ids from the last batch](https://github.com/sgl-project/sglang/blob/392de007cb6f378c9b75546e81f7403b778635e6/python/sglang/srt/managers/scheduler.py#L1752) are simply fake ids that will [become real token](https://github.com/sgl-project/sglang/blob/392de007cb6f378c9b75546e81f7403b778635e6/python/sglang/srt/managers/tp_worker_overlap_thread.py#L170) ids when last batch's computation is finished.
I also don't see how [merging logits processors](https://github.com/sgl-project/sglang/blob/392de007cb6f378c9b75546e81f7403b778635e6/python/sglang/srt/sampling/sampling_batch_info.py#L266) conflicts with the GPU thread.
If you want, there's a [copy_done event](https://github.com/sgl-project/sglang/blob/392de007cb6f378c9b75546e81f7403b778635e6/python/sglang/srt/managers/tp_worker_overlap_thread.py#L207) before fetching the real output_ids from the last batch in the scheduler, which is[ done after every batch forward ](https://github.com/sgl-project/sglang/blob/584e1ab2d06360b3c34a9edbde1679d4bae46c34/python/sglang/srt/managers/scheduler_output_processor_mixin.py#L209)on the CPU overlap thread. You should probably insert that sync [here ](https://github.com/sgl-project/sglang/blob/392de007cb6f378c9b75546e81f7403b778635e6/python/sglang/srt/layers/sampler.py#L263) before applying logits processor.

**Comment by Edenzzzz** [comment]
But still, the logits processor seems to be called inside the sampler on the same thread [as the GPU worker's forward func](https://github.com/sgl-project/sglang/blob/392de007cb6f378c9b75546e81f7403b778635e6/python/sglang/srt/managers/tp_worker_overlap_thread.py#L176), [right after logits computation](https://github.com/sgl-project/sglang/blob/392de007cb6f378c9b75546e81f7403b778635e6/python/sglang/srt/managers/tp_worker.py#L247) , so the last batch's output ids you fetch should have had processor applied, so not sure why that's an issue

<img width="2528" height="800" alt="Image" src="https://github.com/user-attachments/assets/d54a42a9-6a14-4700-be1d-ad6324041a43" />

**Comment by Edenzzzz** [comment]
I see this is because when concatenating `output_ids` of the new batch with the `running_batch` (last batch), this has a race condition with the GPU thread, causing a few tokens in the new batch to not have logits processor applied. 
The best way for now without significant code change is probably to disable the overlap scheduler

<img width="890" height="97" alt="Image" src="https://github.com/user-attachments/assets/6534162a-947a-4d00-87f1-7e50fe0faa4c" />

**Comment by frrad** [comment]
Thanks for looking into this! I will try setting `--disable-overlap-schedule`. 

Maybe I can also add a note about this in the documentation here https://github.com/sgl-project/sglang/blob/21b884606676ac30e3dc0d52ee38e2e5409445ca/docs/basic_usage/sampling_params.md?plain=1#L303-L304

until this bug is resolved.

**Comment by frrad** [comment]
Setting `disable-overlap-schedule` appears to work! Thanks again :)

**Comment by zhaochenyang20** [comment]
@frrad Hey fred, since I think our batch size is not that huge. Overlapped scheduling won't help a lot. 

https://lmsys.org/blog/2024-12-04-sglang-v0-4/#zero-overhead-batch-scheduler
